psi4/driver/broker.py

```diff
@@ -0,0 +1,145 @@
+
+        if molecule is None:
+            molecule = psi4.core.get_active_molecule()
+        self.initial_molecule = molecule
+        assert self.initial_molecule.orientation_fixed() == True, "Orientation must be fixed!"
+        assert self.initial_molecule.point_group().symbol() == "c1", "Symmetry must be 'c1'!"
+
+        names = [self.initial_molecule.symbol(i) for i in range(self.initial_molecule.natom())]
+        psi4.core.print_out("Initial atoms %s\n" % names)
+        self.atoms_list = names
+
+        psi4.core.print_out("PSI4 options:\n")
+        for item, value in self.options.items():
+            psi4.core.print_out("%s %s\n" % (item, value))
+            if item not in ["LOT", "multiplicity", "charge"]:
+                psi4.core.set_global_option(item, value)
+        psi4.core.IO.set_default_namespace("xwrapper")
+
+        self.timing = {}
+
+        atoms = np.array(self.initial_molecule.geometry())
+        psi4.core.print_out("Initial atoms %s\n" % atoms)
+        psi4.core.print_out("Force:\n")
+
+    def calculate_force(self, pos=None, **kwargs):
+        """Fetch force, energy of PSI.
+
+        Arguments:
+        - pos: positions of the atoms as array. If None, the positions of the current active
+          molecule is used.
+        """
+        if pos is None:
+            molecule = psi4.core.get_active_molecule()
+            pos = np.array(molecule.geometry())
+
+        self.frc, self.pot = self.callback(pos, **kwargs)
+        return self.frc, self.pot
+
+    def callback(self, pos, **kwargs):
+        """Initialize psi with new positions and calculate force.
+
+        Arguments:
+        - pos: positions of the atoms as array.
+        """
+
+        self.initial_molecule.set_geometry(psi4.core.Matrix.from_array(pos))
+
+        self.calculate_gradient(self.options["LOT"], **kwargs)
+
+        self.pot = psi4.core.scalar_variable('CURRENT ENERGY')
+        self.frc = -np.array(self.grd)
+
+        return self.frc, np.float64(self.pot)
+
+    def calculate_gradient(self, LOT, bypass_scf=False, **kwargs):
+        """Calculate the gradient with @LOT.
+
+        When bypass_scf=True a hf energy calculation has been done before.
+        """
+        start = time.time()
+        self.grd = psi4.gradient(LOT, bypass_scf=bypass_scf, **kwargs)
+        time_needed = time.time() - start
+        self.timing[LOT] = self.timing.get(LOT, []) + [time_needed]
+
+
+def ipi_broker(molecule=None, serverdata=False, options=None):
+    """ Run IPIBroker to connect to i-pi
+
+    Arguments:
+        molecule: Initial molecule
+        serverdata: Configuration where to connect to ipi
+        options: LOT, multiplicity and charge
+    """
+    b = IPIBroker(molecule=molecule, serverdata=serverdata, options=options)
+
+    try:
+        if b.serverdata:
+            b.run()
+        else:
+            return b
+
+    except KeyboardInterrupt:
+        psi4.core.print_out("Killing IPIBroker\n")
+        b.__del__()
```

Comment:
Usually write this as `del b` and no need to call `sys.exit()` here.

Reply:
I had to call `__del__` explicitly to close the port immediately. with `del b` it seems to be done on the next garbage collection cycle, which does not always seem to happen on termination and the port was still in use after termination. Not sure if this is still required though ..